torchft/process_group.py

```diff
@@ -695,15 +723,18 @@ def _worker(
                 cmd = op[0]
                 if cmd == "func":
                     func_name, args, kwargs = op[1:]
+                    args = _PickleSafeOptions.unsafe_args(args)
                     fn = getattr(pg, func_name)
                     work[next_op_id] = fn(*args, **kwargs)
                     tx.put(next_op_id)
                     next_op_id += 1
                 elif cmd == "wait":
                     op_id: int = op[1]
                     work[op_id].wait()
-                    del work[op_id]
                     tx.put(op_id)
+                elif cmd == "del":
+                    op_id: int = op[1]
+                    del work[op_id]
```

Comment:
Should we wait first before deleting it?

Reply:
I don't think we need to? In normal situations the user code will have waited for it, or they will have gotten a future. If they have a future we still have a reference and don't need it in this map anymore

The only other common case would be if an error occured and we're cleaning things up/shutting down -- in that case waiting likely won't succeed so no reason to wait